### ISSUE TITLE
Remove charset restrictions on baggage item keys

### ISSUE THREAD

**Issue by bensigelman** [opentracing/opentracing.github.io#95]
Per the resurrection of the (dormant) https://github.com/opentracing/opentracing.github.io/issues/25#issuecomment-220827152 , I propose that we get rid of the restrictions on baggage item keys (per SetBaggageItem/GetBaggageItem). Those restrictions were put in place because (a) HTTP is an important and popular inter-process transport, (b) it is appealing to encode individual baggage items as individual HTTP headers, yet (c) HTTP header keys have weird restrictions due to numerous undocumented assumptions and transformations made my numerous distinct HTTP libraries and tools.

Though there's something to be said for the logic above, it feels like the cart is leading the horse. HTTP is not the only transport, nor is it necessary to make the literal baggage item key a component of a corresponding HTTP header key.

**TL;DR:** The proposal is to simply remove all restrictions on the keys (beyond that they be valid unicode strings).

cc some people who've had opinions about this in the past or have dealt with OT implementation: @adriancole @yurishkuro @dkuebric @lookfwd @michaelsembwever @bg451 @bcronin @jmacd @cattail

**Comment by adriancole** [comment]
the impact of this is case preservation becomes an implementation detail
and so folks will have to code accordingly (ex with knowledge that the keys
might not come out the same at the other end).

It might mean http tracers become either more opaque on the wire (being no
longer able to map across headers in order to preserve case (via values)),
or deviant (libraries choosing not to honor the case-sensitivity
constraint).

Regardless, this topic has been well discussed, so nothing more to add.
Thanks for the ping

**Comment by yurishkuro** [comment]
To me, the main drawback is that it complicates the baggage serialization logic in the tracers, especially if we want to keep them backwards compatible with already deployed tracers. But it definitely makes the API easier to use for end users, so I am ok with removing restrictions on baggage keys.

**Comment by dkuebric** [comment]
:+1:

**Comment by bensigelman** [comment]
(I'm making some progress on this during the SpanContext change, but we can't really close this until https://github.com/opentracing/opentracing.github.io/issues/98 is done)

**Comment by bensigelman** [comment]
This is subsumed by https://github.com/opentracing/opentracing.github.io/issues/98 (and the SetBaggageItem comments no longer mention these restrictions in any platform).
